File: leetcode_go/32.Longest_Valid_Parentheses.go
// https: // leetcode.com/problems/longest-valid-parentheses/

func longestValidParentheses(s string) int {
    stack := []int{}
    res := 0
    
    for i, c := range s {
        if c == '(' {
            stack = append(stack, i)
        } else {
            if len(stack) != 0 && s[stack[len(stack)-1]] == '('{
                stack = stack[:len(stack)-1]
                left := -1
                if len(stack) > 0 {
                    left = stack[len(stack)-1]
                }
                if res < i-left {
                    res = i-left
                }
                
            } else {
                stack = append(stack, i)
            }
        }
    }
    return res
}

File: leetcode_java/32.Longest_Valid_Parentheses.java
// https: // leetcode.com/problems/longest-valid-parentheses/

class Solution {
  public int longestValidParentheses(String s) {
    int res = 0;
    Stack<Integer> stack = new Stack<>();

    for (int i = 0; i < s.length(); i++) {
      if (s.charAt(i) == ')') {
        if (!stack.isEmpty() && s.charAt(stack.peek()) == '(') {
          stack.pop();
          int left = stack.isEmpty() ? -1 : stack.peek();
          res = Math.max(res, i - left);
        } else {
          stack.push(i);
        }
      } else {
        stack.push(i);
      }
    }

    return res;
  }
}